### PR TITLE
Communication service database

### PR DESCRIPTION
**Pull request by DesmondTo**
**Updates made:**
1. Chat persistence when the users are just refreshing the page.
2. Change `leave session` button to `finish` as required in the documentation.

> After 5 seconds upon disconnection from the users will cause the others to be kicked out from the room. When no one is in the room, the room will be destroyed and all the chats in the room will be deleted.

https://user-images.githubusercontent.com/70432640/197114259-ee246b5a-4256-4be9-a1f9-79a52a3e2204.mov

**Note:**
- 5 seconds is used to check if the user is just refreshing the page or close the tab for long term. That is, after 5 seconds, if there is only 1 user in the room, the user will be kicked out.

**Issue:**
- For now, if at least one of the users keep refreshing the page for 5 seconds long. Then, both will be kicked out. Not sure if it is a good behaviour since too much refresh will increase the request to the server which adds to additional cost.

### CODE REVIEW

File: collaboration-service/index.js [hunk 1]
```diff
@@ -36,11 +36,26 @@ io.on('connection', (socket) => {
...
-            .emit('session-end', 'Your peer had left the session.', 'warning');
-          io.socketsLeave(room);
+          setTimeout(function () {
+            if (
+              event === sessionEndEvents[0] ||
+              // If after 5 seconds, the room is still less than 2 people,
+              // we can assume that the user had closed the tab.
+              // Otherwise, the user just refreshed the page
+              !io.sockets.adapter.rooms.get(room) ||
+              io.sockets.adapter.rooms.get(room).size < 2
```
Comment (hsiaojietng) [inline collaboration-service/index.js:46]: As mentioned in the meeting, increase to > 5 seconds.

File: collaboration-service/index.js [hunk 2]
```diff
@@ -36,11 +36,26 @@ io.on('connection', (socket) => {
...
+            ) {
+              // Emit to other sockets in the same room this socket had joined
+              socket
+                .to(room)
+                .emit(
+                  'session-end',
+                  'Your peer had left the session.',
```
Comment (hsiaojietng) [inline collaboration-service/index.js:53]: Just a grammar mistake: "Your peer has left the session."
Same below "You have left the session."

File: communication-service/index.js [hunk 3]
```diff
@@ -1,60 +1,48 @@
...
-    if (roomId !== '') {
-      /*
-        When one client send a chat message, every other client(s) 
-        with the same roomId will receive the chat.
-      */
-      socket.to(roomId).emit('receive-chat', senderUsername, chatMessage);
-    }
-  });
+	handleJoinRoom(socket);
+	handleChat(io, socket);
```
Comment (hsiaojietng) [inline communication-service/index.js:32]: Good job in abstracting!

**Review by hsiaojietng — APPROVED** [review]
LGTM, just some nits! Great job!

File: collaboration-service/index.js (hunk 2 above)
Comment (DesmondTo) [reply collaboration-service/index.js]: Oh thanks man, sorry my English is pretty bad haha

File: communication-service/controller/comm-controller.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    };
+  }
+}
+
+export function handleChat(io, socket) {
+  socket.on('send-chat', async (roomId, senderUsername, chatMessage) => {
+    if (roomId !== '') {
+      /*
+			  When one client send a chat message, every other client(s) 
+			  with the same roomId will receive the chat.
```
[the remark this answers is not in the document]
Comment (DesmondTo) [reply communication-service/controller/comm-controller.js]: Not sure why is this happening, but after changing and formatting a few times it came back to normal. Thanks for pointing out 👍

File: collaboration-service/index.js (hunk 1 above)
Comment (DesmondTo) [reply collaboration-service/index.js:46]: Thanks for reminding, I had changed to 10 seconds now.

**Pull request merged**